ibis/backends/duckdb/__init__.py

```diff
@@ -412,9 +415,16 @@ def read_in_memory(
         ir.Table
             The just-registered table
         """
+        import pyarrow as pa
+
         table_name = table_name or util.gen_name("read_in_memory")
         with self.begin() as con:
-            con.connection.register(table_name, dataframe)
+            con.connection.register(table_name, source)
+
+        if isinstance(source, pa.RecordBatchReader):
+            # Ensure the reader isn't marked as started, in case the name is
+            # being overwritten.
+            self._record_batch_reader_states[table_name] = False
```

Comment:
Note that items in this dict are never removed (but they are overwritten). This means that a key-value (`str -> bool`) pair is kept for every unique `RecordBatchReader` name ever queried with a connection. I think this is fine, but wanted to point out the implications. I tried doing some weakref stuff to automatically clear, but duckdb drops the root source `RecordBatchReader` once consumed, so the record that it was already consumed goes away before it's ever referenced to generate the warning.